config/dataset/m2.yaml

```diff
@@ -4,3 +4,4 @@ params:
   n_batches: 10
   batch_size: 20
   num_workers: 0
+  generate_device: "cuda:0"
```

Comment:
Should we specify a particular GPU here? I'm not sure. If it's being committed to our repo, it would seem like the setting would need to be appropriate for anyone who might use these config files. Would an environment variable be more appropriate?

Also, if we're calling it `generate_device`, does the mean the rest of the model might be on a different device? I'm not sure we'd want that ever. If we do use multiple GPUs, it would be in DDP mode (distributed data parallel), so the model would be copied across GPUs.

Reply:
> Should we specify a particular GPU here? I'm not sure. If it's being committed to our repo, it would seem like the setting would need to be appropriate for anyone who might use these config files. Would an environment variable be more appropriate?

Good point, what I've been doing is specify the `gpus` the user wants to use in the [main config file](https://github.com/prob-ml/bliss/blob/master/config/config.yaml), create a `device` variable from this information in the config file (or the command line if the user changes it), and doing something like `torch.cuda.set_device(device)` not sure if there is a better approach.

Reply:
> Should we specify a particular GPU here? I'm not sure. If it's being committed to our repo, it would seem like the setting would need to be appropriate for anyone who might use these config files. Would an environment variable be more appropriate?

I add this in alignment with the `gpus` in the [main config file](https://github.com/prob-ml/bliss/blob/master/config/config.yaml). So I guess the user would need to specify `generate_device` and also `gpus` for the training. But I can make those changes if needed.

> Also, if we're calling it generate_device, does the mean the rest of the model might be on a different device? I'm not sure we'd want that ever. If we do use multiple GPUs, it would be in DDP mode (distributed data parallel), so the model would be copied across GPUs.

Yes, models can be on different devices. This `generate_device` is for hosting the `image_decoder` object in the dataset which is different from the `image_decoder` in our model. This shouldn't break anything as this is similar to how data is moved from cpu to the model gpus before each batch. Here the data is moved from the generating gpu to model gpus. As for whether this would be a good choice for DDP, I'm not sure. Please see the description I just add.

Reply:
I just double-checked and set `generate_device: "cuda:7"`. So the data would be generated on `cuda:7` and the model and training would be on `cuda:0`. It works. `cuda:7` has around 3% workload and around 1100mb memory used, `cuda:0` has around 70% workload and 3300mb memory used. This probably doesn't make much sense, just a sanity check at least that the devices and communication works. I tested "generate_device: cpu" earlier on our server and it also worked.

Reply:
How about we leave the `generate_device` in the main config as optional. When not specified, we just use the first gpu in `gpus`, or cpu if `gpus` is `None`. But it can also be overridden from cli or in config to select a dedicated device for data generation in the event maybe our model saturates gpu memory.